### PR TITLE
fix: test random_port

### LINKED ISSUE #741: add retry mechanism for port conflicts
**Describe the bug** 
The zmq ports conflict in some cases.

```
2020-07-29T16:56:40.0807186Z   File "/home/runner/work/jina/jina/jina/peapods/zmq.py", line 142, in init_sockets
2020-07-29T16:56:40.0807474Z     in_sock, in_addr = _init_socket(ctx, self.args.host_in, self.args.port_in, self.args.socket_in,
2020-07-29T16:56:40.0807935Z   File "/home/runner/work/jina/jina/jina/peapods/zmq.py", line 657, in _init_socket
2020-07-29T16:56:40.0808130Z     raise ex
2020-07-29T16:56:40.0808347Z   File "/home/runner/work/jina/jina/jina/peapods/zmq.py", line 654, in _init_socket
2020-07-29T16:56:40.0808832Z     sock.bind('tcp://%s:%d' % (host, port))
2020-07-29T16:56:40.0809098Z   File "zmq/backend/cython/socket.pyx", line 550, in zmq.backend.cython.socket.Socket.bind
2020-07-29T16:56:40.0809479Z   File "zmq/backend/cython/checkrc.pxd", line 26, in zmq.backend.cython.checkrc._check_rc
2020-07-29T16:56:40.0809669Z zmq.error.ZMQError: Address already in use
```

**Describe how you solve it**

---

**Environment**

**Screenshots**

### PR DESCRIPTION
**Pull request by nan-wang**
resolves #741

### CODE REVIEW

**Comment by hanxiao** [comment]
use draft

**Comment by JoanFM** [comment]
@nan-wang @hanxiao I think the `test_drivers_concat` is broken as is failing for #749

**Comment by hanxiao** [comment]
> @nan-wang @hanxiao I think the `test_drivers_concat` is broken as is failing for #749

image
no problem locally

**Comment by JoanFM** [comment]
> > @nan-wang @hanxiao I think the `test_drivers_concat` is broken as is failing for #749
> 
> image
> no problem locally

Same for me, working locally

**Comment by hanxiao** [comment]
> **WIP, DON'T MERGE**

btw, i've tested both strategies in `random_port`a long time ago and they all collide.

**Comment by nan-wang** [comment]
> > > @nan-wang @hanxiao I think the `test_drivers_concat` is broken as is failing for #749
> > 
> > 
> > image
> > no problem locally
> 
> Same for me, working locally

The last three elements are mismatched.

```
2020-07-31T09:04:00.1119318Z
2020-07-31T09:04:00.1119724Z     def test_direct_concat(self):
2020-07-31T09:04:00.1120299Z         doc1, doc2 = input_fn()
2020-07-31T09:04:00.1120566Z
2020-07-31T09:04:00.1120729Z         t1 = np.concatenate([pb2array(doc1.embedding), pb2array(doc2.embedding)], axis=0)
2020-07-31T09:04:00.1121003Z         doc1.embedding.buffer += doc2.embedding.buffer
2020-07-31T09:04:00.1121344Z         doc1.embedding.shape[0] += doc2.embedding.shape[0]
2020-07-31T09:04:00.1121827Z         t2 = pb2array(doc1.embedding)
2020-07-31T09:04:00.1122012Z
2020-07-31T09:04:00.1122156Z >       np.testing.assert_almost_equal(t1, t2)
2020-07-31T09:04:00.1122302Z E       AssertionError:
2020-07-31T09:04:00.1122676Z E       Arrays are not almost equal to 7 decimals
2020-07-31T09:04:00.1126149Z E
2020-07-31T09:04:00.1126390Z E       Mismatched elements: 3 / 10 (30%)
2020-07-31T09:04:00.1126670Z E       Max absolute difference: 0.28431159
2020-07-31T09:04:00.1126826Z E       Max relative difference: 83.72096832
2020-07-31T09:04:00.1127031Z E        x: array([0.3808869, 0.0888656, 0.8865823, 0.4699178, 0.9115109, 0.0033959,
2020-07-31T09:04:00.1127231Z E              0.1387229, 0.71941  , 0.2877075, 0.7370305])
2020-07-31T09:04:00.1127407Z E        y: array([0.3808869, 0.0888656, 0.8865823, 0.4699178, 0.9115109, 0.0033959,
2020-07-31T09:04:00.1127582Z E              0.1387229, 0.8758986, 0.0033959, 0.9115109])
2020-07-31T09:04:00.1127660Z
2020-07-31T09:04:00.1127811Z tests/unit/drivers/test_concat_driver.py:39: AssertionError
```

**Comment by hanxiao** [comment]
can't reproduce

**Comment by JoanFM** [comment]
I can't reproduce locally, but I guess this byte concatenation must have some issue.

**Comment by hanxiao** [comment]
> I can't reproduce locally, but I guess this byte concatenation must have some issue.

big/little endian things i guess

**Comment by JoanFM** [comment]
From the print output

```bash
tests/unit/drivers/test_concat_driver.py:47: AssertionError
----------------------------- Captured stdout call -----------------------------
[0.5488135  0.71518937 0.60276338 0.54488318 0.4236548  0.64589411 0.43758721]
[0.56804456 0.92559664 0.07103606]

[0.54778474 0.42365479 0.60244746 0.54436832 0.42365479 0.64572212 0.43732048 0.56507891 0.92225854 0.07103606]
[0.54778474 0.42365479 0.60244746 0.54436832 0.42365479 0.64572212 0.43732048 0.5921982  0.71405053 0.42365479]
```

**Comment by hanxiao** [comment]
im confused, what is this PR about?

**Comment by nan-wang** [comment]
> im confused, what is this PR about?

The concate issue is blocking this PR and therefore I just wanted to have a quick look into the issue. This PR is still about fixing the port collision.

**Comment by hanxiao** [comment]
what is the semantic behind `pytest.mark.repeat` and why does it solve our problem? to my understanding on the readme, you are repeating a test until it fails, not repeat a test unit it successes

**Comment by nan-wang** [comment]
> what is the semantic behind `pytest.mark.repeat` and why does it solve our problem? to my understanding of their readme, using `pytest.mark.repeat` means repeating a test until it fails, not repeating until it successes.

`pytest.mark.repeat` does **NOT** solve the problem. It is used to repeat the test for multiple times and ensure it doesn't pop up randomly. 

The problem is solved by setting `JINA_RANDOM_PORTS`. We've already been using `pytest-xdist` and set `-n 1` to ensure only one test process is running. Therefore the port collision must due to the conflicts within each test. In each test, we ask `socket` to assign port for us and quickly free it. The `socket` will reuse the freed socket and assigned to another Pod in the same test.

This solution is not elegant. I'm working on another solution.

**Comment by nan-wang** [comment]
## Problems
The reason behind the port collision is as discussed previously. In short, **the port assigned by `socket` in the `random_port()` is possible been taken by other usages before `zmq` initialization**. There are two cases lead to the this problem.
1. Collision within each flow. 
2. Collision with other process. 

## Insights
The first case can be fixed by keeping track of all the ports used by the Pods in the same flow.

As for the second case, it happens more often on the server, where the port are open and closed frequently for various purposes. To fix this, we need to assign a port during `zmq` initialization when a collision happens. Depends on the type of ports that collide, we need different way to send a `port-change` message to other Pods/clients so that they can modify the ports accordingly. 

| port type | what to update and how |
| --- | --- |
|`port_input` | only the Pod itself |
|`port_output` | all the Pods listening to the port|
|`port_control` | both the Pod itself and the client |
|`port_expose` | both the Pod itself and the client |


## Status-quo
The first case is solved by this PR. In the second case, I'm working on the updates of the Pod itself. We still miss a way to send the `port-change` message to the other Pods/clients.

**Comment by hanxiao** [comment]
when touching the flow bear in mind that pod can be local /remote and inside docker

File: jina/helper.py [hunk 1]
```diff
@@ -260,20 +260,33 @@ def random_name() -> str:
...
-        from contextlib import closing
-        import socket
         import threading
         with threading.Lock():
             with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
                 s.bind(('', 0))
                 s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-                return s.getsockname()[1]
+                _port = s.getsockname()[1]
+                return _port
```
Comment (hanxiao) [inline jina/helper.py:273]:
```suggestion
                _port = s.getsockname()[1]
                return _port
```
```suggestion
                return s.getsockname()[1]
```

File: jina/helper.py [hunk 2]
```diff
@@ -260,20 +260,33 @@ def random_name() -> str:
...
+            with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+                result = s.connect_ex(('', p))
+                s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+                return result == 0
+
         min_port, max_port = 49152, 65535
-        return random.randrange(min_port, max_port)
+        _port = None
+        while is_port_in_use(_port):
+            _port = random.randrange(min_port, max_port)
```
Comment (hanxiao) [inline jina/helper.py:288]: the coding style needs to be improved.

can't you set `_port = random.randrange(min_port, max_port)` first and then do the `while`? this remove `if p is None return True`

File: jina/peapods/zmq.py [hunk 3]
```diff
@@ -316,15 +316,18 @@ def close(self):
...
+                # Replace handle events function, to skip
+                # None event after sockets are closed.
+                if hasattr(self.in_sock, '_handle_events'):
+                    self.in_sock._handle_events = lambda *args, **kwargs: None
+                if hasattr(self.out_sock, '_handle_events'):
+                    self.out_sock._handle_events = lambda *args, **kwargs: None
+                if hasattr(self.ctrl_sock, '_handle_events'):
+                    self.ctrl_sock._handle_events = lambda *args, **kwargs: None
+            except AttributeError as e:
+                self.logger.info(f'failed to stop. {e}')
```
Comment (hanxiao) [inline jina/peapods/zmq.py:330]: what could possibly happen here?

Comment (nan-wang) [reply jina/peapods/zmq.py:330]: When the address conflict happens, these attributes will not be set and this leads to an exception  in closing the process and messed up the error messages.

> what could possibly happen here?

When the address conflict happens, these attributes will not be set and this leads to an exception  in closing the process and messed up the error messages.

Comment (hanxiao) [reply jina/peapods/zmq.py:330]: then it should be error not info

File: jina/flow/__init__.py [hunk 4]
```diff
@@ -834,3 +835,18 @@ def use_grpc_gateway(self):
...
+    def _check_port_collision(self, kwargs):
+        """Check if the Pods' ports collide"""
+        for _port_name in ('port_in', 'port_out', 'port_ctrl', 'port_expose'):
+            _port = kwargs.get(_port_name)
+            if _port is None:
+                _port = random_port()
+            while _port in self._ports_in_use:
+                _new = random_port()
+                _port = _new
+                self.logger.info(f'{_port_name} collision detected. set from {_port} to {_new}')
```
Comment (hanxiao) [inline jina/flow/__init__.py:848]: this should be warning

File: jina/flow/__init__.py [hunk 5]
```diff
@@ -834,3 +835,18 @@ def use_grpc_gateway(self):
...
+            kwargs[_port_name] = _port
+            self._ports_in_use.append(_port)
+            self.logger.info(f'{kwargs.get("name")} {_port_name}: {_port}')
```
Comment (hanxiao) [inline jina/flow/__init__.py:851]: to debug

File: jina/helper.py [hunk 6]
```diff
@@ -272,8 +272,19 @@ def random_port() -> int:
                 return s.getsockname()[1]
     else:
         import random
+
+        def is_port_in_use(p):
```
Comment (hanxiao) [inline jina/helper.py:276]: repeated code, no threading lock

File: jina/flow/__init__.py [hunk 7]
```diff
@@ -398,8 +399,8 @@ def add(self,
 
         kwargs.update(op_flow._common_kwargs)
         kwargs['name'] = pod_name
+        op_flow._ports_in_use += self._check_port_collision(kwargs)
```
Comment (hanxiao) [inline jina/flow/__init__.py:402]: you can not use `self` here, it is a factory pattern, use `op_flow`

**Pull request merged**